### PR TITLE
simple fix for empty lastobs_df when streamnudging unknowingly turned on while no usgs data is available

### PR DESCRIPTION
**Pull request by kumdonoaa**
This fix provides a solution when lastobs_df is empty because no usgs data is available while stream nudging is unknowingly activated in the config file. More solid solution will be provided soon.  Also, in diffusive_routing mainstem_list now has a list  of stream segments aligned in order from upstream to downstream.  Moreover, missing topobathy data for a stream segment get filled in by looking up the newly aligned list.

## Additionns
**AbstractRouting.py**
- removed sort() to mainstem_list to align a list of stream segments in order from upstream to downstream
- modified the way to search for an upstream segment id for a given mainstem segment that has no channel xsec topobathy data

**DataAssimilation.py**
- Added a condition not to call new_lastobs function when there is no lastobs related values updated from mc_reach.pyx

**output.py**
- Added a condition not to call lastobs_df_output when there is no lastobs related values updated from mc_reach.pyx

## Removals

-

## Changes

-

## Testing

1.

## Notes

-

## Todos

-

## Checklist

- [ ] PR has an informative and human-readable title
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code can be automatically merged (no conflicts)
- [ ] Code follows project standards (link if applicable)
- [ ] Passes all existing automated tests
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future todos are captured in comments
- [ ] Visually tested in supported browsers and devices (see checklist below :point_down:)
- [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [ ] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Target Environment support

- [ ] Windows
- [ ] Linux
- [ ] Browser

### Accessibility

- [ ] Keyboard friendly
- [ ] Screen reader friendly

### Other

- [ ] Is useable without CSS
- [ ] Is useable without JS
- [ ] Flexible from small to large screens
- [ ] No linting errors or warnings
- [ ] JavaScript tests are passing

### CODE REVIEW

File: src/troute-network/troute/DataAssimilation.py [hunk 1]
```diff
@@ -238,10 +238,12 @@ def update_after_compute(self, run_results, time_increment):
             - lastobs_df               (DataFrame): Last gage observations data for DA
         '''
         streamflow_da_parameters = self._data_assimilation_parameters.get('streamflow_da', None)
- 
+
         if streamflow_da_parameters:
             if streamflow_da_parameters.get('streamflow_nudging', False):
-                self._last_obs_df = new_lastobs(run_results, time_increment)
+                if not run_results[0][3][0].size == 0:
```
Comment (shorvath-noaa) [inline src/troute-network/troute/DataAssimilation.py:244]: I think this check needs to be modified. This only checks if the first object in `run_results` has no gage information. But this varies for each object:

```
(Pdb) [rr[3][0].size for rr in run_results]
[2,2,0,2,0,1,0,0,3,0,17,1,1,4,3,1,0,0,0,2,0,1,2,0,0,0,1,0,3,1,2,1,1,0,1,1,0,1,0,1,3,0,0,0,0,0,0,0,3,1,0,11,0,2,0,0,3,1,0,0,0,2,4,0,0,3,0,0,0,0]
(Pdb)
```

File: src/troute-network/troute/AbstractRouting.py [hunk 2]
```diff
@@ -312,10 +312,19 @@ def topobathy_df(self):
...
-                            upstream_mainstem_segment = set(self._diffusive_network_data[tw]['rconn'][mainstem_segment]).intersection(
-                                                                                    self._diffusive_network_data[tw]['mainstem_segs'])
-                            upstream_mainstem_segment = list(upstream_mainstem_segment)[0]
-                            temp_df = self._topobathy_df[self._topobathy_df.index==upstream_mainstem_segment]
+                            temp_df = pd.DataFrame()
+                            position_mainstem_segment =  self._diffusive_network_data[tw]['mainstem_segs'].index(mainstem_segment)
+                            position_upstream_mainstem_segment = position_mainstem_segment
+                            while temp_df.empty:               
+                                try:
+                                    position_upstream_mainstem_segment = position_upstream_mainstem_segment - 1
```
Comment (AminTorabi-NOAA) [inline src/troute-network/troute/AbstractRouting.py:320]: Why do you decrease it by 1 here? Are the IDs always consecutive?

File: src/troute-network/troute/DataAssimilation.py (hunk 1 above)
Comment (kumdonoaa) [reply src/troute-network/troute/DataAssimilation.py:244]: what about checking self._last_obs_df? If it is empty in the first place, then we expect nothing will come out for lastobs values from mc_reach.pyx.

File: src/troute-network/troute/AbstractRouting.py (hunk 2 above)
Comment (kumdonoaa) [reply src/troute-network/troute/AbstractRouting.py:320]: That does not decreasing flowpath ID number but the position of stream segment in the mainstem segment list that has aligned the list from upstream to downstream.

Comment (AminTorabi-NOAA) [reply src/troute-network/troute/AbstractRouting.py:320]: True that's correct. Just a very small suggestion, its more readable to write it like `position_upstream_mainstem_segment -= 1`

File: src/troute-network/troute/DataAssimilation.py (hunk 1 above)
Comment (kumdonoaa) [reply src/troute-network/troute/DataAssimilation.py]: Thanks for your suggestion.

File: src/troute-network/troute/AbstractRouting.py (hunk 2 above)
Comment (kumdonoaa) [reply src/troute-network/troute/AbstractRouting.py]: Thanks and updated.

**Pull request merged**